File: airesources/Clojure/src/hlt/networking.clj
(ns hlt.networking
  (:require
   [clojure.string :as str]
   [hlt.entity :as e]
   [hlt.math :as math]))

(defn read-prelude
  [{:keys [in]}]
  (let [player-id (read in)
        width (read in)
        height (read in)]
    {:player-id player-id
     :map-size [width height]}))

;; decoding game map
(defn- to-docking-status
  [docking-status]
  (case docking-status
    0 :undocked
    1 :docking
    2 :docked
    3 :undocking))

(defn- parse-ship
  [stream owner-id]
  (let [ship-id          (read stream)
        x-loc            (read stream)
        y-loc            (read stream)
        health           (read stream)
        x-velocity       (read stream) ;; velocity, which is deprecated
        y-velocity       (read stream)
        docking-status   (to-docking-status (read stream))
        docked-planet    (read stream)
        docking-progress (read stream)
        weapon-cooldown  (read stream)] ;; weapon cooldown seems unused as well
    (e/->Ship ship-id (math/->Position x-loc y-loc) health e/ship-radius owner-id
              {:status docking-status
               :planet docked-planet
               :progress docking-progress})))

(defn- parse-ship-list
  [stream owner-id]
  (let [num-ships (read stream)]
    (loop [ships (transient {})
           cur-ship 0]
      (if (== num-ships cur-ship)
        (persistent! ships)
        (let [ship (parse-ship stream owner-id)]
          (recur (assoc! ships (:id ship) ship)
                 (inc cur-ship)))))))

(defn- parse-all-ships
  [stream]
  (let [num-players (read stream)]
     (loop [all-ships (transient {})
            owner-ships {}
            cur-player 0]
       (if (== num-players cur-player)
         {:owner-ships owner-ships
          :ships (persistent! all-ships)}
         (let [owner-id (read stream)
               ships (parse-ship-list stream owner-id)]
           (recur (reduce conj! all-ships ships)
                  (assoc owner-ships owner-id ships)
                  (inc cur-player)))))))

(defn- parse-planet
  [stream]
  (let [id                   (read stream)
        x-loc                (read stream)
        y-loc                (read stream)
        health               (read stream)
        radius               (read stream)
        docking-spots        (read stream)
        current-production   (read stream) ;; Deprecated
        remaining-production (read stream) ;;
        has-owner            (read stream)
        owner-candidate      (read stream)
        docked-ship-count    (read stream)
        docked-ships (vec (repeatedly docked-ship-count #(read stream)))]
    (e/->Planet id (math/->Position x-loc y-loc) health radius
                (if (== has-owner 1)
                  owner-candidate
                  nil)
                {:spots docking-spots
                 :ships docked-ships})))

(defn- parse-all-planets
  [stream]
  (let [num-planets (read stream)]
    (loop [planets (transient {})
           cur-planet 0]
      (if (== num-planets cur-planet)
        (persistent! planets)
        (let [planet (parse-planet stream)]
          (recur (assoc! planets (:id planet) planet)
                 (inc cur-planet)))))))

(defn read-map
  [{:keys [in]}]
  (let [ships   (parse-all-ships in)
        planets (parse-all-planets in)]
    (assoc ships :planets planets)))

;; finish initialization

(defn send-done-initialized
  "Notifies the game engine that this bot has been initialized."
  [{:keys [out]} bot-name]
  (doto out
    (.write (str bot-name "\n"))
    (.flush)))

;; encoding moves

(def ^:private undock-key "u")
(def ^:private dock-key "d")
(def ^:private thrust-key "t")

(defmulti move-segments :type)

(defmethod move-segments :undock [move]
  [undock-key (-> move :ship :id)])

(defmethod move-segments :dock [move]
  [dock-key (-> move :ship :id) (-> move :planet :id)])

(defmethod move-segments :thrust [move]
  [thrust-key (-> move :ship :id) (:thrust move) (:angle move)])

(defn send-moves
  "Sends the moves to the game engine."
  [{:keys [out]} moves & r]
  (let [move-seq (mapcat move-segments moves)]
    (doto out
      (.write (str (str/join " " move-seq) "\n"))
      (.flush))))

File: airesources/Haskell/Hlt/Networking.hs
-- | Communicate with the Halite game environment.
module Hlt.Networking where

import qualified Data.Map as Map
import Text.Read
import System.IO
import Hlt.Constants
import Hlt.Entity
import Hlt.GameMap

instance Read PlayerId where
    readPrec = do
      a <- step readPrec :: ReadPrec Int
      return $ PlayerId a

instance Read Player where
    readPrec = do
        i <- step readPrec :: ReadPrec PlayerId -- id
        n <- step readPrec :: ReadPrec Int      -- number of ships
        let next c = if c == 0 then
                         return []
                     else do
                         s <- step readPrec :: ReadPrec Ship
                         ss <- next (c-1)
                         return $ (shipId s, s):ss
        s <- next n                             -- list of ships
        return $ Player i (Map.fromList s)

instance Read ShipId where
    readPrec = do
        a <- step readPrec :: ReadPrec Int
        return $ ShipId a

instance Read DockingStatus where
    readPrec = do
        a <- step readPrec :: ReadPrec Int
        return $ case a of
            0 -> Undocked
            1 -> Docking
            2 -> Docked
            3 -> Undocking

instance Read Ship where
    readPrec = do
        i <- step readPrec :: ReadPrec ShipId        -- id
        j <- step readPrec :: ReadPrec Float         -- x
        k <- step readPrec :: ReadPrec Float         -- y
        h <- step readPrec :: ReadPrec Int           -- health
        _ <- step readPrec :: ReadPrec Float         -- x velocity (deprecated)
        _ <- step readPrec :: ReadPrec Float         -- y velocity (deprecated)
        s <- step readPrec :: ReadPrec DockingStatus -- docking status
        p <- step readPrec :: ReadPrec PlanetId      -- docked planet
        d <- step readPrec :: ReadPrec Int           -- docking progress
        c <- step readPrec :: ReadPrec Int           -- weapon cooldown
        return $ Ship i j k h s (if s /= Undocked then Just p else Nothing) d c

instance Read PlanetId where
    readPrec = do
        a <- step readPrec :: ReadPrec Int
        return $ PlanetId a

instance Read Planet where
    readPrec = do
        i <- step readPrec :: ReadPrec PlanetId -- id
        j <- step readPrec :: ReadPrec Float    -- x
        k <- step readPrec :: ReadPrec Float    -- y
        h <- step readPrec :: ReadPrec Int      -- health
        r <- step readPrec :: ReadPrec Float    -- radius
        d <- step readPrec :: ReadPrec Int      -- docking spots
        p <- step readPrec :: ReadPrec Int      -- production
        _ <- step readPrec :: ReadPrec Int      -- remaining production (deprecated)
        w <- step readPrec :: ReadPrec Int      -- whether it is owned
        o <- step readPrec :: ReadPrec PlayerId -- the player id if it is owned
        s <- step readPrec :: ReadPrec Int      -- number of docked ships
        let next c = if c == 0 then return []
                     else do
                        s <- step readPrec :: ReadPrec ShipId
                        ss <- next (c-1)
                        return $ s:ss
        l <- next s                             -- list if ship ids
        return $ Planet i j k h r d p (if w == 1 then Just o else Nothing) l

-- | Read a list of Players and Ships from a string,
readPlayers :: Int -> String -> (([(PlayerId, Player)], String))
readPlayers 0 s = ([], s)
readPlayers c s = do
    let (a, r0) = head $ (reads :: ReadS Player) s
        (b, r1) = readPlayers (c-1) r0
    ((playerId a, a):b, r1)

-- | Read a list of Planets from a string,
readPlanets :: Int -> String -> (([(PlanetId, Planet)], String))
readPlanets 0 s = ([], s)
readPlanets c s = do
    let (a, r0) = head $ (reads :: ReadS Planet) s
        (b, r1) = readPlanets (c-1) r0
    ((planetId a, a):b, r1)

-- | Get the next int and return the rest of the string.
nextInt :: String -> (Int, String)
nextInt s = head $ (reads :: ReadS Int) s

-- | Read players and planets from stdin.
readGameMap :: IO (Map.Map PlayerId Player,  Map.Map PlanetId Planet)
readGameMap = do
    line <- getLine
    let (c0, r0) = nextInt line
        (p0, r1) = readPlayers c0 r0
        (c1, r2) = nextInt r1
        (p1, r3) = readPlanets c1 r2
    return (Map.fromList p0, Map.fromList p1)

-- | Send a string to the game environment.
sendString :: String -> IO ()
sendString s = do
    putStrLn s
    hFlush stdout

-- | Send a list of commands to the game environment.
sendCommands :: [String] -> IO ()
sendCommands = sendString . unwords

-- | Get the initial GameMap.
initialGameMap :: IO GameMap
initialGameMap = do
    line0 <- getLine -- our player id
    line1 <- getLine -- game dimensions
    let i = read line0 :: PlayerId
        (w, h) = (\[a, b] -> (a, b)) $ map read $ words line1
    (p0, p1) <- readGameMap
    return (GameMap i w h p0 p1)

-- | Get the new GameMap.
updateGameMap :: GameMap -> IO GameMap
updateGameMap g = do
    (p0, p1) <- readGameMap
    return (GameMap (myId g) (width g) (height g) p0 p1)
